sdk/schemaregistry/azure-data-schemaregistry-jsonschema/src/main/java/com/azure/data/schemaregistry/jsonschema/SchemaRegistrySchemaCache.java

```diff
@@ -79,16 +149,61 @@ Mono<String> getSchemaId(String schemaFullName, String schemaDefinition) {
         });
     }
 
-    Mono<String> getSchema(String schemaId) {
+    /**
+     * The schema id.  {@link SchemaRegistryAsyncClient} is used if {@link SchemaRegistryClient} is not set.
+     *
+     * @param schemaId Id to get the schema definition.
+     * @return The schema definition associated with the id.
+     */
+    String getSchemaDefinition(String schemaId) {
+        if (schemaRegistryClient == null) {
+            return getSchemaDefinitionAsync(schemaId).block();
+        }
+
+        final String existing;
         synchronized (lock) {
-            final String existing = cache.get(schemaId);
+            existing = cache.get(schemaId);
+        }
 
-            if (existing != null) {
-                return Mono.just(existing);
-            }
+        if (existing != null) {
+            return existing;
+        }
+
+        final SchemaRegistrySchema registryObject = schemaRegistryClient.getSchema(schemaId);
+        final String schemaString = registryObject.getDefinition();
+
+        synchronized (lock) {
+            cache.put(schemaId, schemaString);
+            logCacheStatus();
+        }
+
+        return schemaString;
+    }
+
+    /**
+     * The schema id.  {@link SchemaRegistryAsyncClient} is used if {@link SchemaRegistryClient} is not set.
+     *
+     * @param schemaId Id to get the schema definition.
+     * @return The schema definition associated with the id.
+     * @throws IllegalStateException When {@link SchemaRegistryAsyncClient} is not set.
+     */
+    Mono<String> getSchemaDefinitionAsync(String schemaId) {
+        final String existing;
+        synchronized (lock) {
+            existing = cache.get(schemaId);
+        }
+
+        if (existing != null) {
+            return Mono.just(existing);
+        }
+
+        if (Objects.isNull(schemaRegistryAsyncClient)) {
+            return monoError(LOGGER, new IllegalStateException("Cannot use async methods if SchemaRegistryAsyncClient "
+                + "Set client in SchemaRegistryJsonSchemaSerializer.schemaRegistryClient(SchemaRegistryAsyncClient)."
+            ));
```

Comment:
Maybe a bad design idea from me but having a mysterious can't use this because early configuration wasn't defined seems bad. Something like this may warrant fully splitting sync and async into different clients to ensure correct usage.

Reply:
Sadly, we decided to go with 1 client.

Reply:
Yeah, if we had not used single serializer for both sync and async in Avro, we could have split this. However, it's not too bad as users know that they are using sync client and they don't want to switch to async calls. The error is clear enough to indicate what is missing. If we do get customer feedback to support this, we can `Mono.defer(() -> callSyncClient())` later.